contrib/mesos/pkg/election unit test timeout

https://travis-ci.org/GoogleCloudPlatform/kubernetes/jobs/69776608

```
panic: test timed out after 5m0s
goroutine 37 [running]:
runtime.panic(0xaef900, 0xc2080aa020)
	/usr/local/go/src/pkg/runtime/panic.c:279 +0xf5
testing.func·008()
	/usr/local/go/src/pkg/testing/testing.go:629 +0x12e
created by time.goFunc
	/usr/local/go/src/pkg/time/sleep.go:121 +0x55
goroutine 16 [chan receive, 2 minutes]:
testing.RunTests(0xee2130, 0x13b34a0, 0x4, 0x4, 0xe3fe01)
	/usr/local/go/src/pkg/testing/testing.go:505 +0xbb8
testing.Main(0xee2130, 0x13b34a0, 0x4, 0x4, 0x13e7900, 0x0, 0x0, 0x13e7900, 0x0, 0x0)
	/usr/local/go/src/pkg/testing/testing.go:435 +0xa3
main.main()
	github.com/GoogleCloudPlatform/kubernetes/contrib/mesos/pkg/election/_test/_testmain.go:105 +0x18c
goroutine 19 [finalizer wait, 5 minutes]:
runtime.park(0x435ff0, 0x13de610, 0x13b8349)
	/usr/local/go/src/pkg/runtime/proc.c:1369 +0x89
runtime.parkunlock(0x13de610, 0x13b8349)
	/usr/local/go/src/pkg/runtime/proc.c:1385 +0x3b
runfinq()
	/usr/local/go/src/pkg/runtime/mgc0.c:2644 +0xcf
runtime.goexit()
	/usr/local/go/src/pkg/runtime/proc.c:1445
goroutine 33 [chan receive]:
github.com/golang/glog.(*loggingT).flushDaemon(0x13e1280)
	/home/travis/gopath/src/github.com/GoogleCloudPlatform/kubernetes/Godeps/_workspace/src/github.com/golang/glog/glog.go:879 +0x96
created by github.com/golang/glog.init·1
	/home/travis/gopath/src/github.com/GoogleCloudPlatform/kubernetes/Godeps/_workspace/src/github.com/golang/glog/glog.go:410 +0x2d6
goroutine 17 [syscall, 5 minutes]:
runtime.goexit()
	/usr/local/go/src/pkg/runtime/proc.c:1445
goroutine 52 [select, 2 minutes]:
github.com/GoogleCloudPlatform/kubernetes/contrib/mesos/pkg/election.(*etcdMasterElector).run(0xc20804d1a0, 0xd24260, 0x3, 0xd21a40, 0x3)
	github.com/GoogleCloudPlatform/kubernetes/contrib/mesos/pkg/election/_test/_obj_test/etcd_master.go:49 +0x3db
github.com/GoogleCloudPlatform/kubernetes/contrib/mesos/pkg/election.func·001()
	github.com/GoogleCloudPlatform/kubernetes/contrib/mesos/pkg/election/_test/_obj_test/etcd_master.go:37 +0xb9
github.com/GoogleCloudPlatform/kubernetes/pkg/util.func·004()
	/home/travis/gopath/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/pkg/util/util.go:112 +0x5f
github.com/GoogleCloudPlatform/kubernetes/pkg/util.Until(0xc20804d1c0, 0x12a05f200, 0x0)
	/home/travis/gopath/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/pkg/util/util.go:113 +0xbc
github.com/GoogleCloudPlatform/kubernetes/pkg/util.Forever(0xc20804d1c0, 0x12a05f200)
	/home/travis/gopath/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/pkg/util/util.go:94 +0x48
created by github.com/GoogleCloudPlatform/kubernetes/contrib/mesos/pkg/election.(*etcdMasterElector).Elect
	github.com/GoogleCloudPlatform/kubernetes/contrib/mesos/pkg/election/_test/_obj_test/etcd_master.go:37 +0x1f6
goroutine 49 [select, 2 minutes]:
github.com/GoogleCloudPlatform/kubernetes/contrib/mesos/pkg/election.(*etcdMasterElector).run(0xc20804d0e0, 0xd24260, 0x3, 0xd21a40, 0x3)
	github.com/GoogleCloudPlatform/kubernetes/contrib/mesos/pkg/election/_test/_obj_test/etcd_master.go:49 +0x3db
github.com/GoogleCloudPlatform/kubernetes/contrib/mesos/pkg/election.func·001()
	github.com/GoogleCloudPlatform/kubernetes/contrib/mesos/pkg/election/_test/_obj_test/etcd_master.go:37 +0xb9
github.com/GoogleCloudPlatform/kubernetes/pkg/util.func·004()
	/home/travis/gopath/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/pkg/util/util.go:112 +0x5f
github.com/GoogleCloudPlatform/kubernetes/pkg/util.Until(0xc20804d100, 0x12a05f200, 0x0)
	/home/travis/gopath/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/pkg/util/util.go:113 +0xbc
github.com/GoogleCloudPlatform/kubernetes/pkg/util.Forever(0xc20804d100, 0x12a05f200)
	/home/travis/gopath/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/pkg/util/util.go:94 +0x48
created by github.com/GoogleCloudPlatform/kubernetes/contrib/mesos/pkg/election.(*etcdMasterElector).Elect
	github.com/GoogleCloudPlatform/kubernetes/contrib/mesos/pkg/election/_test/_obj_test/etcd_master.go:37 +0x1f6
goroutine 55 [select, 2 minutes]:
github.com/GoogleCloudPlatform/kubernetes/contrib/mesos/pkg/election.(*etcdMasterElector).run(0xc20804d320, 0xd24260, 0x3, 0xd21a40, 0x3)
	github.com/GoogleCloudPlatform/kubernetes/contrib/mesos/pkg/election/_test/_obj_test/etcd_master.go:49 +0x3db
github.com/GoogleCloudPlatform/kubernetes/contrib/mesos/pkg/election.func·001()
	github.com/GoogleCloudPlatform/kubernetes/contrib/mesos/pkg/election/_test/_obj_test/etcd_master.go:37 +0xb9
github.com/GoogleCloudPlatform/kubernetes/pkg/util.func·004()
	/home/travis/gopath/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/pkg/util/util.go:112 +0x5f
github.com/GoogleCloudPlatform/kubernetes/pkg/util.Until(0xc20804d340, 0x12a05f200, 0x0)
	/home/travis/gopath/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/pkg/util/util.go:113 +0xbc
github.com/GoogleCloudPlatform/kubernetes/pkg/util.Forever(0xc20804d340, 0x12a05f200)
	/home/travis/gopath/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/pkg/util/util.go:94 +0x48
created by github.com/GoogleCloudPlatform/kubernetes/contrib/mesos/pkg/election.(*etcdMasterElector).Elect
	github.com/GoogleCloudPlatform/kubernetes/contrib/mesos/pkg/election/_test/_obj_test/etcd_master.go:37 +0x1f6
goroutine 58 [select, 2 minutes]:
github.com/GoogleCloudPlatform/kubernetes/pkg/watch.(*Broadcaster).distribute(0xc2080b1110, 0xd372d0, 0x8, 0x7f1745fd6ef0, 0xc208132d20)
	/home/travis/gopath/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/pkg/watch/mux.go:182 +0x3bd
github.com/GoogleCloudPlatform/kubernetes/pkg/watch.(*Broadcaster).loop(0xc2080b1110)
	/home/travis/gopath/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/pkg/watch/mux.go:163 +0xc3
created by github.com/GoogleCloudPlatform/kubernetes/pkg/watch.NewBroadcaster
	/home/travis/gopath/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/pkg/watch/mux.go:70 +0x127
goroutine 57 [chan receive, 2 minutes]:
github.com/GoogleCloudPlatform/kubernetes/contrib/mesos/pkg/election.Test(0xc2080a6750)
	/home/travis/gopath/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/contrib/mesos/pkg/election/master_test.go:83 +0x306
testing.tRunner(0xc2080a6750, 0x13b34e8)
	/usr/local/go/src/pkg/testing/testing.go:422 +0x110
created by testing.RunTests
	/usr/local/go/src/pkg/testing/testing.go:504 +0xb46
goroutine 59 [select, 2 minutes]:
github.com/GoogleCloudPlatform/kubernetes/contrib/mesos/pkg/election.(*notifier).serviceLoop(0xc208093b30, 0xc2080977a0)
	github.com/GoogleCloudPlatform/kubernetes/contrib/mesos/pkg/election/_test/_obj_test/master.go:105 +0x363
github.com/GoogleCloudPlatform/kubernetes/contrib/mesos/pkg/election.func·005()
	github.com/GoogleCloudPlatform/kubernetes/contrib/mesos/pkg/election/_test/_obj_test/master.go:87 +0x90
github.com/GoogleCloudPlatform/kubernetes/contrib/mesos/pkg/runtime.func·008()
	/home/travis/gopath/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/contrib/mesos/pkg/runtime/util.go:115 +0x5f
github.com/GoogleCloudPlatform/kubernetes/contrib/mesos/pkg/runtime.Until(0x7f174580bec0, 0x0, 0xc2080976e0)
	/home/travis/gopath/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/contrib/mesos/pkg/runtime/util.go:116 +0xd3
github.com/GoogleCloudPlatform/kubernetes/contrib/mesos/pkg/election.Notify(0x7f1745fd7030, 0xc20804d460, 0xd16620, 0x0, 0xd26520, 0x2, 0x7f1745fd7058, 0xc20804d480, 0xc2080976e0)
	github.com/GoogleCloudPlatform/kubernetes/contrib/mesos/pkg/election/_test/_obj_test/master.go:87 +0x429
github.com/GoogleCloudPlatform/kubernetes/contrib/mesos/pkg/election.func·006()
	/home/travis/gopath/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/contrib/mesos/pkg/election/master_test.go:72 +0x145
github.com/GoogleCloudPlatform/kubernetes/contrib/mesos/pkg/runtime.func·007()
	/home/travis/gopath/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/contrib/mesos/pkg/runtime/util.go:95 +0xaa
created by github.com/GoogleCloudPlatform/kubernetes/contrib/mesos/pkg/runtime.After
	/home/travis/gopath/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/contrib/mesos/pkg/runtime/util.go:97 +0xf7
goroutine 61 [chan send, 2 minutes]:
github.com/GoogleCloudPlatform/kubernetes/contrib/mesos/pkg/election.(*Fake).Elect(0xc20804d460, 0xd16620, 0x0, 0xd26520, 0x2, 0x0, 0x0)
	github.com/GoogleCloudPlatform/kubernetes/contrib/mesos/pkg/election/_test/_obj_test/fake.go:38 +0x23e
github.com/GoogleCloudPlatform/kubernetes/contrib/mesos/pkg/election.func·003()
	github.com/GoogleCloudPlatform/kubernetes/contrib/mesos/pkg/election/_test/_obj_test/master.go:45 +0x13c
github.com/GoogleCloudPlatform/kubernetes/contrib/mesos/pkg/runtime.func·008()
	/home/travis/gopath/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/contrib/mesos/pkg/runtime/util.go:115 +0x5f
github.com/GoogleCloudPlatform/kubernetes/contrib/mesos/pkg/runtime.Until(0x7f1745807f50, 0x0, 0xc2080976e0)
	/home/travis/gopath/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/contrib/mesos/pkg/runtime/util.go:116 +0xd3
github.com/GoogleCloudPlatform/kubernetes/contrib/mesos/pkg/election.func·004()
	github.com/GoogleCloudPlatform/kubernetes/contrib/mesos/pkg/election/_test/_obj_test/master.go:84 +0x14f
github.com/GoogleCloudPlatform/kubernetes/contrib/mesos/pkg/runtime.func·007()
	/home/travis/gopath/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/contrib/mesos/pkg/runtime/util.go:95 +0xaa
created by github.com/GoogleCloudPlatform/kubernetes/contrib/mesos/pkg/runtime.After
	/home/travis/gopath/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/contrib/mesos/pkg/runtime/util.go:97 +0xf7
goroutine 62 [semacquire, 2 minutes]:
sync.runtime_Syncsemacquire(0xc2080a54d0)
	/usr/local/go/src/pkg/runtime/sema.goc:257 +0xc0
sync.(*Cond).Wait(0xc2080a54c0)
	/usr/local/go/src/pkg/sync/cond.go:62 +0xb5
sync.*Cond.Wait·fm()
	github.com/GoogleCloudPlatform/kubernetes/contrib/mesos/pkg/election/_test/_obj_test/master.go:104 +0x34
github.com/GoogleCloudPlatform/kubernetes/contrib/mesos/pkg/runtime.func·007()
	/home/travis/gopath/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/contrib/mesos/pkg/runtime/util.go:95 +0xaa
created by github.com/GoogleCloudPlatform/kubernetes/contrib/mesos/pkg/runtime.After
	/home/travis/gopath/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/contrib/mesos/pkg/runtime/util.go:97 +0xf7
FAIL	github.com/GoogleCloudPlatform/kubernetes/contrib/mesos/pkg/election	300.121s
```

/cc @jdef, @davidopp

--- a/contrib/mesos/pkg/election/master.go
+++ b/contrib/mesos/pkg/election/master.go
@@ -46,14 +46,14 @@ type Service interface {
 }
 
 type notifier struct {
-	lock sync.Mutex
-	cond *sync.Cond
+	changed chan struct{} // to notify the service loop about changed state
 
 	// desired is updated with every change, current is updated after
 	// Start()/Stop() finishes. 'cond' is used to signal that a change
 	// might be needed. This handles the case where mastership flops
 	// around without calling Start()/Stop() excessively.
 	desired, current Master
+	lock             sync.Mutex // to protect the desired variable
 
 	// for comparison, to see if we are master.
 	id Master
@@ -65,7 +65,7 @@ type notifier struct {
 // elected master starts/stops matching 'id'. Never returns.
 func Notify(m MasterElector, path, id string, s Service, abort <-chan struct{}) {
 	n := &notifier{id: Master(id), service: s}
-	n.cond = sync.NewCond(&n.lock)
+	n.changed = make(chan struct{})
 	finished := runtime.After(func() {
 		runtime.Until(func() {
 			for {
@@ -86,14 +86,16 @@ func Notify(m MasterElector, path, id string, s Service, abort <-chan struct{}) 
 							glog.Errorf("Unexpected object from election channel: %v", event.Object)
 							break
 						}
-						func() {
-							n.lock.Lock()
-							defer n.lock.Unlock()
-							n.desired = electedMaster
-							if n.desired != n.current {
-								n.cond.Signal()
-							}
-						}()
+
+						n.lock.Lock()
+						n.desired = electedMaster
+						n.lock.Unlock()
+
+						// notify serviceLoop, but don't block. If a change
+						// is queued already it will see the new n.desired.
+						select {
+						case n.changed <- struct{}{}:
+						}
 					}
 				}
 			}
@@ -104,31 +106,22 @@ func Notify(m MasterElector, path, id string, s Service, abort <-chan struct{}) 
 
 // serviceLoop waits for changes, and calls Start()/Stop() as needed.
 func (n *notifier) serviceLoop(abort <-chan struct{}) {
-	n.lock.Lock()
-	defer n.lock.Unlock()
 	for {
 		select {
 		case <-abort:
 			return
-		default:
-			for n.desired == n.current {
-				ch := runtime.After(n.cond.Wait)
-				select {
-				case <-abort:
-					n.cond.Signal() // ensure that Wait() returns
-					<-ch
-					return
-				case <-ch:
-					// we were notified and have the lock, proceed..
-				}
-			}
-			if n.current != n.id && n.desired == n.id {
-				n.service.Validate(n.desired, n.current)
+		case <-n.changed:
+			n.lock.Lock()
+			newDesired := n.desired // copy value to avoid race below
+			n.lock.Unlock()
+
+			if n.current != n.id && newDesired == n.id {
+				n.service.Validate(newDesired, n.current)
 				n.service.Start()
-			} else if n.current == n.id && n.desired != n.id {
+			} else if n.current == n.id && newDesired != n.id {
 				n.service.Stop()
 			}
-			n.current = n.desired
+			n.current = newDesired
 		}
 	}
 }

--- a/contrib/mesos/pkg/election/master_test.go
+++ b/contrib/mesos/pkg/election/master_test.go
@@ -69,7 +69,23 @@ func Test(t *testing.T) {
 	changes := make(chan bool, 1500)
 	done := make(chan struct{})
 	s := &slowService{t: t, changes: changes, done: done}
+
+	// change master to "notme" such that the initial m.Elect call inside Notify
+	// will trigger an obversable event. We will wait for it to make sure the
+	// Notify loop will see those master changes triggered by the go routine below.
+	m.ChangeMaster(Master("me"))
+	temporaryWatch := m.mux.Watch()
+	ch := temporaryWatch.ResultChan()
+
 	notifyDone := runtime.After(func() { Notify(m, "", "me", s, done) })
+
+	// wait for the event triggered by the initial m.Elect of Notify. Then drain
+	// the channel to not block anything.
+	<-ch
+	temporaryWatch.Stop()
+	for i := 0; i < len(ch); i += 1 { // go 1.3 and 1.4 compatible loop
+		<-ch
+	}
 
 	go func() {
 		defer close(done)
@@ -83,16 +99,8 @@ func Test(t *testing.T) {
 	<-notifyDone
 	close(changes)
 
-	changeList := []bool{}
-	for {
-		change, ok := <-changes
-		if !ok {
-			break
-		}
-		changeList = append(changeList, change)
-	}
-
-	if len(changeList) > 1000 {
-		t.Errorf("unexpected number of changes: %v", len(changeList))
+	changesNum := len(changes)
+	if changesNum > 1000 || changesNum == 0 {
+		t.Errorf("unexpected number of changes: %v", changesNum)
 	}
 }

